src/layouts/AppContainer.tsx

```diff
@@ -64,59 +66,88 @@ class AppContainer extends React.Component<IProps, IState> {
+     */
+    let currentAddress = await getCurrentAccountAddress() || null;
+    let accountWasCached = false;
     if (currentAddress)  {
-      console.log(`using address from web3 connection: ${currentAddress}`);
-      localStorage.setItem(storageKey, currentAddress);
+      console.log(`using account from existing web3Provider: ${currentAddress}`);
+      this.cacheWeb3Info(currentAddress);
     } else {
-      currentAddress = localStorage.getItem(storageKey);
+      /**
+       * Heads up that there is a chance this cached account may differ from an account
+       * that the user has already selected in a provider but have
+       * not yet made available to the app (if the account were available we'd have caught it
+       * with getCurrentAccountAddress).
+       */
+      currentAddress = this.getCachedAccount();
       if (currentAddress) {
-        console.log(`using address from local storage: ${currentAddress}`);
-      } else {
-        localStorage.setItem(storageKey, "");
+        accountWasCached = true;
+        console.log(`using account from local storage: ${currentAddress}`);
       }
     }
 
     this.props.setCurrentAccount(currentAddress);
+    /**
+     * Only supply currentAddress if it was obtained from a provider.  The poll
+     * is only comparing changes with respect to the provider state.  Passing it a cached state
+     * will only cause it to get the wrong impression and misbehave.
+     */
+    pollForAccountChanges(accountWasCached ? null : currentAddress).subscribe(
+      (newAddress: Address | null): void => {
+        console.log(`new account: ${newAddress}`);
+        this.props.setCurrentAccount(newAddress);
+        if (newAddress) {
+          this.cacheWeb3Info(newAddress);
+        } else {
+          this.uncacheWeb3Info();
+        }
+      });
+  }
 
-    try {
-      metamask = await checkWeb3Provider();
-    } catch (err) {
-      console.log("MM not available or not set correctly: using default web3 provider: ", err.message);
-    }
-
-    if (metamask) {
-      pollForAccountChanges(currentAddress).subscribe(
-        (newAddress: Address) => {
-          if (newAddress && checkWeb3Provider()) {
-            this.props.setCurrentAccount(newAddress);
-            localStorage.setItem(storageKey, newAddress);
-
-            initializeArc();
-          }
-        });
+  private async loadCachedWeb3Provider(): Promise<boolean> {
+    const web3ProviderInfo = this.getCachedWeb3ProviderInfo();
+    if (web3ProviderInfo) {
+      /**
+       * If successful, this will result in setting the current account which
+       * we'll pick up below.
+       */
+      if (await setWeb3Provider(web3ProviderInfo)) {
+        console.log("using cached web3Provider");
+        return true;
+      } else {
+        console.log("failed to instantiate cached web3Provider");
+        this.uncacheWeb3Info();
+      }
     }
+    return false;
   }
 
-  public render() {
+  public render(): any {
```

Comment:
yeah im over this lint rule

Reply:
Not me, at all.  'any' is really not a big deal.  Not having the the non-trivial return values is.

Reply:
I think if both @jellegerbrandy and @tibetsprague prefer not to have this rule we should turn it off @dkent600.